src/lib/Imp.hs

```diff
@@ -94,7 +94,7 @@ toImpCExpr dests op = case op of
     n' <- typeToSize (varAnn b)
     emitLoop n' $ \i -> do
       let ithDest = tabGetIAtom dests i
-      extendR (b @> asIdx n' i) $ toImpExpr ithDest body
+      extendR (b @> asIdx (varAnn b) i) $ toImpExpr ithDest body
```

Comment:
**Important note:** This assumes that the `AsIdx` constructor always expects expressions in the range [0, n) where n is the size of the index space (even though the real index space might be e.g. `Range 5 10`).